typing_inspect/typing_inspect.py

```diff
@@ -0,0 +1,248 @@
+"""Defines experimental API for runtime inspection of types defined
+in the standard "typing" module.
+
+Example usage::
+    from typing_inspect import is_generic
+"""
+
+# NOTE: This module must support Python 2.7 in addition to Python 3.x
+
+from typing import Callable, Union, Tuple, TypeVar, _gorg, ClassVar, Generic
+
+
+def is_generic_type(tp):
+    """Test if the given type is a generic type. This includes Generic itself, but
+    excludes special typing constructs such as Union, Tuple, Callable, ClassVar.
+    Examples::
+
+        is_generic_type(int) == False
+        is_generic_type(Union[int, str]) == False
+        is_generic_type(Union[int, T]) == False
+        is_generic_type(ClassVar[List[int]]) == False
+        is_generic_type(Callable[..., T]) == False
+
+        is_generic_type(Generic) == True
+        is_generic_type(Generic[T]) == True
+        is_generic_type(Iterable[int]) == True
+        is_generic_type(Mapping) == True
+        is_generic_type(MutableMapping[T, List[int]]) == True
+        is_generic_type(Sequence[Union[str, bytes]]) == True
+    """
+
+    return (isinstance(tp, type(Generic)) and not
+            isinstance(tp, (type(Callable), type(Tuple))))
+
+
+def is_callable_type(tp):
+    """Test if the type is a generic callable type, including subclasses
+    excluding non-generic types and callables.
+    Examples::
+
+        is_callable_type(int) == False
+        is_callable_type(type) == False
+        is_callable_type(Callable) == True
+        is_callable_type(Callable[..., int]) == True
+        is_callable_type(Callable[[int, int], Iterable[str]]) == True
+        class MyClass(Callable[[int], int]):
+            ...
+        is_callable_type(MyClass) == True
+
+    For more general tests use callable(), for more precise test use::
+
+        get_origin(tp) is Callable
```

Comment:
What is the difference exactly between this construct and `is_callable_type`? I think this could be explained more.